kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -484,6 +484,178 @@ sealed abstract class Resource[F[_], +A] {
     new (F ~> F) {
       override def apply[B](gb: F[B]): F[B] = surround(gb)
     }
+
+  def forceR[B](that: Resource[F, B])(implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
+    Resource applyFull { poll =>
+      poll(this.use_ !> that.allocated) map { p =>
+        // map exists on tuple in Scala 3 and has the wrong type
+        Functor[(B, *)].map(p)(fin => (_: Resource.ExitCase) => fin)
+      }
+    }
+
+  def onCancel[B >: A](fin: Resource[F, Unit])(
+      implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
+    Resource applyFull { poll =>
+      poll(this.allocated).onCancel(fin.use_) map { p =>
+        Functor[(B, *)].map(p)(fin => (_: Resource.ExitCase) => fin)
+      }
+    }
+
+  def guaranteeCase[B >: A](fin: Outcome[Resource[F, *], Throwable, B] => Resource[F, Unit])(
+      implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
```

Comment:
```suggestion
  def guaranteeCase(fin: Outcome[Resource[F, *], Throwable, A @uncheckedVariance] => Resource[F, Unit])(
      implicit F: MonadCancel[F, Throwable]): Resource[F, A] =
```

This is one case where I believe `@uncheckedVariance` is actually needed, for the same reason it is needed in `IO`.